casr/src/triage.rs

```diff
@@ -0,0 +1,464 @@
+
+        if casr_cluster_c.status.success() {
+            info!(
+                "{}",
+                String::from_utf8_lossy(&casr_cluster_c.stdout).trim_end()
+            );
+        } else {
+            error!("{}", String::from_utf8_lossy(&casr_cluster_c.stderr));
+        }
+
+        // Remove reports from deduplication phase. They are in clusters now.
+        for casrep in fs::read_dir(output_dir)?.flatten().map(|e| e.path()) {
+            if let Some(ext) = casrep.extension() {
+                if ext == "casrep" {
+                    let _ = fs::remove_file(casrep);
+                }
+            }
+        }
+    }
+
+    summarize_results(matches, crashes)
+}
+
+/// Copy crashes next to reports and print summary.
+/// Run casr-gdb on uninstrumented binary if specified in ARGS.
+/// Print analysis statistic.
+///
+/// # Arguments
+///
+/// * `matches` - tool arguments
+///
+/// * `crashes` - set of crashes, specified as a CrashInfo structure
+fn summarize_results(
+    matches: &clap::ArgMatches,
+    crashes: &HashMap<String, CrashInfo>,
+) -> Result<()> {
+    // Get output dir
+    let dir = matches.get_one::<PathBuf>("output").unwrap();
+    // Copy crashes next to reports
+    copy_crashes(dir, crashes)?;
+
+    // Get casr-gdb args
+    let tool = std::env::current_exe()?
+        .file_name()
+        .unwrap()
+        .to_str()
+        .unwrap()
+        .to_string();
+    let gdb_args = if tool.eq("casr-libfuzzer") {
+        // casr-libfuzzer case
+        if let Some(argv) = matches.get_one::<String>("casr-gdb-args") {
+            shell_words::split(argv)?
+        } else {
+            Vec::new()
+        }
+    } else if let Some(argv) = matches.get_many::<String>("ARGS") {
+        // casr-afl case
+        argv.cloned().collect()
+    } else {
+        Vec::new()
+    };
+    // Get timeout
+    let timeout = if let Some(timeout) = matches.get_one::<u64>("timeout") {
+        *timeout
+    } else {
+        0
+    };
+    // Get number of threads
+    let jobs = if let Some(jobs) = matches.get_one::<u32>("jobs") {
+        *jobs as usize
+    } else {
+        std::cmp::max(1, num_cpus::get() / 2)
+    };
+
+    if !gdb_args.is_empty() {
+        let casr_gdb = get_path("casr-gdb")?;
+        // Run casr-gdb on uninstrumented binary.
+        let crashes: Vec<_> = WalkDir::new(dir)
+            .into_iter()
+            .filter_entry(|e| {
+                let name = e.file_name().to_str().unwrap();
+                !name.eq("clerr") && !name.eq("oom") && !name.eq("timeout")
+            })
+            .flatten()
+            .map(|e| e.into_path())
+            .filter(|e| e.is_file())
+            .filter(|e| e.extension().is_none() || e.extension().unwrap() != "casrep")
+            .filter(|e| !Path::new(format!("{}.gdb.casrep", e.display()).as_str()).exists())
+            .collect();
+        let num_of_threads = jobs.min(crashes.len()).max(1) + 1;
```

Comment:
we should check that num_of_threads > 1

We should not log anything and generate casr reports if there are no crashes

Reply:
Now, we check that crashes are not empty